src/ApiHandler.ts

```diff
@@ -644,4 +739,114 @@ export default class ApiHandler {
 			args: newArgs,
 		};
 	}
+
+	/**
+	 * Derive information on the progress of the current session and era.
+	 *
+	 * @param api ApiPromise with ensured metadata.
+	 * @param hash block hash to get info at.
+	 */
+	private async deriveSessionAndEraProgress(
+		api: ApiPromise,
+		hash: BlockHash
+	): Promise<{
+		eraLength: BN;
+		eraProgress: BN;
+		sessionLength: BN;
+		sessionProgress: BN;
+		activeEra: EraIndex;
+	}> {
+		const [
+			currentSlot,
+			epochIndex,
+			genesisSlot,
+			currentIndex,
+			activeEraOption,
+		] = await Promise.all([
+			await api.query.babe.currentSlot.at(hash),
+			await api.query.babe.epochIndex.at(hash),
+			await api.query.babe.genesisSlot.at(hash),
+			await api.query.session.currentIndex.at(hash),
+			await api.query.staking.activeEra.at(hash),
+		]);
+
+		const { index: activeEra } = activeEraOption.unwrapOrDefault();
+
+		// TODO make sure this doesn't fail on querys older than history depth
```

Comment:
it shouldn't: active era is (due to current pallet-session) at most one behind "current era" (considered as current era from the point of view of pallet-staking)
Thus active start will never be less than `current_era-history_depth`

Reply:
Could active-era be an arbitrary amount behind current-era?

Reply:
pallet-session only delay by one session so active-era is at least current-era - 1.

But in general if we change pallet session (I mean its code) and make it delaying by 3 sessions and new_era is trigger always then yes, active-era will be 3 index behind current-era.